Sources/Extensions/UIKit/UIRefreshControlExtensions.swift

```diff
@@ -0,0 +1,29 @@
+//
+//  UIRefreshControlExtensions.swift
+//  SwifterSwift
+//
+//  Created by ratul sharker on 7/24/18.
+//  Copyright © 2018 SwifterSwift
+//
+
+#if canImport(UIKit)
+import UIKit
+
+#if !os(watchOS)
+// MARK: - Methods
+public extension UIRefreshControl {
+
+    /// SwifterSwift: Programatically begin refresh control inside of UITableView.
+    ///
+    /// - Parameter tableView: UITableView instance, inside which the refresh control is contained.
+    public func beginRefreshProgramatically(in tableView: UITableView) {
+        //https://stackoverflow.com/questions/14718850/uirefreshcontrol-beginrefreshing-not-working-when-uitableviewcontroller-is-ins/14719658#14719658
+        beginRefreshing()
+        let offsetPoint = CGPoint(x: 0, y: -frame.height)
+        tableView.setContentOffset(offsetPoint, animated: true)
```

Comment:
This isn't enough to actually get it to work. To register for when a pull-to-refresh action has been called on a `UIRefreshControl` you need to add a target-action for `.valueChanged`. You can instantly call
`sendActions(for: .valueChanged)`
but then you won't see the animation for showing the refresh control. Taking into account this, and the comments mentioned above, I think that the final version should be:

```Swift
extension UITableView {
    public func refresh(animated: Bool = true) {
        let refreshControl: UIRefreshControl!
        if #available(iOS 10.0, *) {
            refreshControl = self.refreshControl
        } else {
            refreshControl = subviews.first(where: { $0 is UIRefreshControl }) as? UIRefreshControl
        }

        guard refreshControl != nil else { return }

        // https://stackoverflow.com/questions/14718850/uirefreshcontrol-beginrefreshing-not-working-when-uitableviewcontroller-is-ins/14719658#14719658
        refreshControl.beginRefreshing()
        let offsetPoint = CGPoint(x: 0, y: -refreshControl.frame.height)
        setContentOffset(offsetPoint, animated: animated)

        DispatchQueue.main.asyncAfter(deadline: .now() + (animated ? 0.3 : 0.0)) { [weak refreshControl] in
            refreshControl?.sendActions(for: .valueChanged)
        }
    }
}
```
I created a test project and it looks like `0.3` is the time used for `setContentOffset(_:, animated:)`. It may not be ideal to use it since it can change in future versions of iOS, but Apple doesn't provide us the constant (that I'm aware of) and we can't use the delegate callback `scrollViewDidEndScrollingAnimation(_:)` here.

Please also note the function name in my version. Maybe you can come up with a better one, but `beginRefreshProgramatically` is both a mouthful and it's not very "clean" to add `programmatically` (there are 2 m's, not 1) to the name of a programming function.

Reply:
If you want to avoid calling `sendActions` after the delay, the only alternative I can think of is to add a big note in the comments that the caller needs to ensure that the delegate of the table view calls the target-action callback manually once `scrollViewDidEndScrollingAnimation(_:)` was called, but that's very clunky.

Reply:
<h2>1. Calling with wrong `UITableView` as parameter</h2>
This could be easily checked is the developer called with wrong TV. RC is added as a subview of the TV. Below iOS 10 developer deliberately add RC as subview of TV, so it is obvious. For iOS 10+ i have tested as follows

```
func testExample() {
    let tableView = UITableView.init()
    tableView.refreshControl = UIRefreshControl.init()
    XCTAssertEqual(tableView, tableView.refreshControl?.superview)
}
```

The test passes. So if we put a checking in the extension like following

```
public func beginRefreshProgramatically(in tableView: UITableView) {
       if superview == tableView {
            beginRefreshing()
            let offsetPoint = CGPoint(x: 0, y: -frame.height)
            tableView.setContentOffset(offsetPoint, animated: true)
        } else {
            print("Refresh Control does not belong to this table view")
            // we can also put assertFailure here, but i'm unsure about it
        }
    }
```
The problem of setting content offset of the wrong TV can be avoided like this.

<h2>2. Supporting Pre iOS 10 Solution</h2>
In iOS 10+ RC can be added as subview in TV, and they just work fine just like pre iOS 10. The developer could've coded like this even after his app targeting iOS 10+. Adding the RC as subview won't set the TV's `refreshControl` property. i.e

```
func testTableViewAddSubView() {
        let tableView = UITableView.init()
        let refreshControl = UIRefreshControl.init()
        tableView.addSubview(refreshControl)
        XCTAssertNotEqual(tableView.refreshControl, refreshControl)
    }
```
In this case the solution won't work because the checking is put like following

```
if #available(iOS 10.0, *) {
    refreshControl = self.refreshControl
}
```
which won't get the RC object and the RC won't refresh anything.

More worse case could be like following. Think of an application which support pre iOS 10 using the solution you provided and the application is running on a device with iOS 10+. As the application supports pre iOS 10 then RC is added as subview, not set in the property. As your proposed solution the code should reach Line 2 of the following block.

```
1. if #available(iOS 10.0, *) {
2.    refreshControl = self.refreshControl
3. }
```
As the RC is added as subview in TV it won't get the `refreshControl` object, resulting in setting the nil into the `refreshControl` and ended up crashing while calling `beginRefreshing()` method or accessing the `frame` property.

<h2>3. Why the extension still in UIRefreshControl class</h2>
UIRefreshControl widget is responsible for showing refreshing animation, not the UITableView. It is more likely that, developer who using SwifterSwift will find refresh method in UIRefreshControl object. <b>Even after adding the `refreshControl` property in iOS 10+ you have to call `beginRefreshing` method on the RC not in the TV</b> . This is more developer friendly to put this extension into the RC instead of TV. You could not assume that the developer will read the SwifterSwift documentation and find this magic inside TV. Even i don't read the SwifterSwift documentation because i don't have to. All the extension i guessed, found them in the right place.

<h2>4. To see the animation the delay of 0.3</h2>
As you said apple does not provide any constant of the delays, i strongly disagree to add this portion in the extension. A try to move this extension from RC to TV is the obvious reason for this. This is another reason i don't want this extension in RC. I didn't think so far, but thanks for your finding. You may have ran your test project in few devices. Have you ran them on all devices with different iOS version running on them ? In past i had very difficult time for taking such decision, like introduce a constant delay to sync with some tasks. If you like you can lookup `DEFAULT_KEYBOARD_UP_ANIMATION_DURATION` 
<a href="https://github.com/ratulSharker/Keyboard-Animator/blob/master/Example%20Project/Testing/Testing/KeyboardAnimator/KeyboardAnimator.h">here</a> . For each and every iOS version i ended up tweaking this constant value. I could find a value which seems perfect for all iOS and devices. Later i came to know the key `UIKeyboardAnimationDurationUserInfoKey` which solves all the problem. The alternative to using constant duration very unfriendly api (letting the developer call `scrollViewDidEndScrollingAnimation` manually through a delegate ) and the delegate you are asking as a alternative is too much to use this functionality. It just an overkill.

<h2>5. Naming convention</h2>
Yes, in the name i missed two `m`. My mistake. The name may sound too mouthful, but i wanted to keep the `beingRefresh` as a prefix. The reason is while the developer will begin typing like

```
refreshControl.beg
```
I want xcode suggest the native `beginRefresh` and the extension side by side. Please let me know if you can suggest something less mouthful and cleaner which is prefixed with `beginRefresh` or i should go with `beginRefreshProgramatically` / `beginRefreshProgrammatically`.

Let me know your suggestion regarding these 5 statements.

Reply:
1. I agree that we can easily create an assert for it.
2. You're right, the user might have added the RC as a subview so my check would have failed (it wouldn't have crashed because of the `guard`, though). That's easily fixed.
3. The user will be calling the function the same way. They won't need to call `beginRefreshing` themselves if they use the TV extension, but they do still need that target-action callback that calls `refreshControl.endRefreshing`, so I see your point that it makes sense to keep them together.
4. Yeah, the animation timing is definitely a bad idea. If the reloading of the table view is instantaneous, which in real world apps is unlikely, then they just won't see the refresh control, it'll scroll back up to the top and the user can keep using the table view instantly.
5. I think it's fine to just use `beginRefreshing` for the name. The user will see the difference between the original and the extension because of the parameters.

Based on your points, I suggest this extension:
```Swift
extension UIRefreshControl {
    public func beginRefreshing(in tableView: UITableView, animated: Bool = true) {
        let isSubView = {
            tableView.subviews.contains(where: { $0 == self })
        }
        if #available(iOS 10.0, *) {
            assert(isSubView() || tableView.refreshControl == self, "Refresh control does not belong to the receiving table view")
        } else {
            assert(isSubView(), "Refresh control does not belong to the receiving table view")
        }

        beginRefreshing()
        let offsetPoint = CGPoint(x: 0, y: -frame.height)
        tableView.setContentOffset(offsetPoint, animated: animated)

        sendActions(for: .valueChanged)
    }
}
```
`isSubview` is a closure because on non-debug builds (where `assert` isn't executed) we don't want to actually perform the calculation. Chances are that the number of subviews will be small, but why risk a potentially heavy calculation?

It works perfectly in my test VC:

```Swift
class ViewController: UIViewController, UITableViewDataSource {
    @IBOutlet weak var tableView: UITableView!

    var refreshControl: UIRefreshControl!

    override func viewDidLoad() {
        super.viewDidLoad()

        refreshControl = UIRefreshControl(frame: CGRect(origin: .zero, size: CGSize(width: view.frame.width, height: 44)))
        refreshControl.addTarget(self, action: #selector(refresh(_:)), for: [.valueChanged])
        if #available(iOS 10.0, *) {
            tableView.refreshControl = refreshControl
        } else {
            tableView.addSubview(refreshControl)
        }
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        refreshControl.beginRefreshing(in: tableView)
    }

    @objc func refresh(_ sender: UIRefreshControl) {
        // Wait to show the RC
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { [weak self] in
            self?.refreshControl.endRefreshing()
        }
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 500
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        cell.textLabel?.text = indexPath.description
        return cell
    }
}
```

Thoughts?

Reply:
In the suggested extension 

```
let isSubView = {
            tableView.subviews.contains(where: { $0 == self })
        }
        if #available(iOS 10.0, *) {
            assert(isSubView() || tableView.refreshControl == self, "Refresh control does not belong to the receiving table view")
        } else {
            assert(isSubView(), "Refresh control does not belong to the receiving table view")
        }
```

It finds the RC from the TV. Why not just check the `superview` of the RC ?

```
if superview == tableView {
        // valid
} else {
       // invalid -- assert here
}
```

It also survives the iOS 10+, as `refreshControl` property in TV is added as subview. i.e 

```
func testIsRefreshControlAddedAsSubview(){
        let tableView = UITableView.init()
        tableView.refreshControl = UIRefreshControl.init()
        XCTAssertEqual(tableView, tableView.refreshControl?.superview)
    }
```

which is passed.

This superview checking version is technically more efficient, manageable and most of all it supports both pre and post iOS 10.

I really appreciate the animation parameter. 👍 

But the `sendActions(for: .valueChanged)` i totally disagree with. As you can see, while changing state of the `UISwitch`, `UISegmentController`, `UIProgressView` and others, none of them call the added target while developer change the state of the widget. 

Let me clarify my point, say you added a target and selector to the `UISegmentSelector`. While the user taps the segment controller and select the different segment the selector of that target is getting called. But while developer programmatically set or change the selected index, the selector in the target not getting called. 

This is more likely that developer will expect the selector in the target won't get called. Most of the time developer want to change the state because the job required to do while the state changes are already done, thats why changing the state deliberately by the system. So changing the state does not mean to fire up the selector attached to it. But we can keep it with another parameter deciding to fire it up or not.

So summarizing everything, Here is a version that i suggest to keep.

```
public func beginRefreshing(in tableView: UITableView, animated: Bool, sendAction: Bool) {
        //https://stackoverflow.com/questions/14718850/uirefreshcontrol-beginrefreshing-not-working-when-uitableviewcontroller-is-ins/14719658#14719658
        assert(superview == tableView, "Refresh control does not belong to the receiving table view")
        
        beginRefreshing()
        let offsetPoint = CGPoint(x: 0, y: -frame.height)
        tableView.setContentOffset(offsetPoint, animated: animated)

        if sendAction {
            sendActions(for: .valueChanged)
        }
    }
```

Please share your thought about it.

Reply:
I totally agree with everything.
I was sure that I tested the `superview` value for iOS 10 and it came back as `false`, which is why I did both checks. I tested it again and you're right, it's `true`.
I don't even think we need the `sendAction` parameter. As you said, it would make it more in line with other `UIControl`s. Also, if the developer is already calling this method they know that they need to perform the reloading, they don't need to wait for the target-action callback. This makes a lot of sense.
Thanks for putting up with my debate :)
💯